x/tally/keeper/filter.go

```diff
@@ -28,69 +71,42 @@ type FilterResult struct {
 // index i is an outlier, consensus boolean, consensus data proxy
 // public keys, and error. It assumes that the reveals and their
 // proxy public keys are sorted.
-func (k Keeper) ApplyFilter(ctx sdk.Context, input []byte, reveals []types.RevealBody, replicationFactor uint16) (FilterResult, error) {
+func ApplyFilter(filter types.Filter, reveals []types.RevealBody) (FilterResult, error) {
 	var result FilterResult
+	result.Errors = make([]bool, len(reveals))
 	result.Outliers = make([]int, len(reveals))
+	result.GasUsed = filter.GasCost()
 
-	if len(input) == 0 {
-		return result, types.ErrInvalidFilterType
-	}
-
-	// Determine basic consensus on tuple of (exit_code, proxy_pub_keys)
+	// Determine basic consensus on tuple of (exit_code_success, proxy_pub_keys)
 	var maxFreq int
-	var proxyPubKeys []string
 	freq := make(map[string]int, len(reveals))
-	for _, reveal := range reveals {
+	for i, reveal := range reveals {
 		success := reveal.ExitCode == 0
+		result.Errors[i] = !success
 		tuple := fmt.Sprintf("%v%v", success, reveal.ProxyPubKeys)
 		freq[tuple]++
 
 		if freq[tuple] > maxFreq {
-			proxyPubKeys = reveal.ProxyPubKeys
+			result.ProxyPubKeys = reveal.ProxyPubKeys
 			maxFreq = freq[tuple]
 		}
 	}
-	if maxFreq*3 < len(reveals)*2 {
+	if maxFreq*3 < len(reveals)*2 { // TODO remove basic consensus check?
 		return result, types.ErrNoBasicConsensus
 	}
```

Comment:
I think it makes sense to keep this short-circuit in. Makes the rest of the filter code much easier to read.